aspnetcore/security/authentication/certauth.md

```diff
@@ -206,3 +206,249 @@ Complete the following steps In IIS Manager:
+
+### Using certificate authentication with the X-ARR-ClientCert header
+
+The AddCertificateForwarding method is used so that the client header can be specified and how the certificate is to be loaded using the HeaderConverter option. When sending the certificate with the HttpClient using the default settings, the ClientCertificate was always be null. The X-ARR-ClientCert header is used to pass the client certificate, and the cert is passed as a string to work around this.
+
+```csharp
+services.AddCertificateForwarding(options =>
+{
+	options.CertificateHeader = "X-ARR-ClientCert";
+	options.HeaderConverter = (headerValue) =>
+	{
+		X509Certificate2 clientCertificate = null;
+		if(!string.IsNullOrWhiteSpace(headerValue))
+		{
+			byte[] bytes = StringToByteArray(headerValue);
+			clientCertificate = new X509Certificate2(bytes);
+		}
+
+		return clientCertificate;
+	};
+});
+```
+
+The Configure method then adds the middleware. UseCertificateForwarding is added before the UseAuthentication and the UseAuthorization.
+
+```csharp
+public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
+{
+	...
+	
+	app.UseRouting();
+
+	app.UseCertificateForwarding();
+	app.UseAuthentication();
+	app.UseAuthorization();
+
+	app.UseEndpoints(endpoints =>
+	{
+		endpoints.MapControllers();
+	});
+}
+```
+
+A separate class can be used to implement validation logic. Because we are using the same self signed certificate in this example, we need to ensure that only our certificate can be used. We validate that the thumbprints of the client certificate and also the server one match, otherwise any certificate can be used and will be be enough to authenticate. This would be used inside the AddCertificate method. You could also validate the subject or the issuer here, if you are using intermediate or child certificates.
+```csharp
+using System.IO;
+using System.Security.Cryptography.X509Certificates;
+
+namespace AspNetCoreCertificateAuthApi
+{
+    public class MyCertificateValidationService
+    {
+        public bool ValidateCertificate(X509Certificate2 clientCertificate)
+        {
+            var cert = new X509Certificate2(Path.Combine("sts_dev_cert.pfx"), "1234");
+            if (clientCertificate.Thumbprint == cert.Thumbprint)
+            {
+                return true;
+            }
+
+            return false;
+        }
+    }
+}
+
+```
+
+#### Implementing a HttpClient using a certificate
+
+The client of the API uses a HttpClient which was create using an instance of the IHttpClientFactory. This does not provide a way to define a handler for the HttpClient and so we use a HttpRequestMessage to add the Certificate to the "X-ARR-ClientCert" request header. The cert is added as a string using the GetRawCertDataString method. 
+ 
+```csharp
+private async Task<JArray> GetApiDataAsync()
```

Comment:
Bonus points if we can use something in `System.Text.Json` instead Json.NET's `JArray` type.

Reply:
switched to JsonDocument